cli/src/peers.rs

```diff
@@ -0,0 +1,81 @@
+// Finding a peer at most once, so we can build urls.
+
+use anyhow::{anyhow, Result};
+use async_once_cell::OnceCell;
+use tokio::time::sleep;
+use utils::mesh::{KaboodleMesh, PeerMetadata, ServalMesh, ServalRole};
+
+use std::{net::SocketAddr, time::Duration};
+
+static SERVAL_NODE_ADDR: OnceCell<SocketAddr> = async_once_cell::OnceCell::new();
+
+async fn base_url() -> SocketAddr {
+    *SERVAL_NODE_ADDR
+        .get_or_init(async {
+            maybe_find_peer(None, "SERVAL_NODE_URL")
+                .await
+                .expect("unable to find any mesh peers!")
+        })
+        .await
+}
+
+// Convenience function to build urls repeatably.
+pub async fn build_url(path: String, version: Option<&str>) -> String {
+    let baseurl = base_url().await;
+    if let Some(v) = version {
+        format!("http://{baseurl}/v{v}/{path}")
+    } else {
+        format!("http://{baseurl}/{path}")
+    }
+}
+
+pub async fn create_mesh_peer() -> Result<ServalMesh> {
+    let host = std::env::var("HOST").unwrap_or_else(|_| "0.0.0.0".to_string());
+    let (port, interface) = utils::mesh::mesh_interface_and_port();
+
+    let metadata = PeerMetadata::new(
+        format!("client@{host}"),
+        None,
+        vec![ServalRole::Client],
+        None,
+    );
+    let mut mesh = ServalMesh::new(metadata, interface, port).await?;
+    mesh.start().await?;
+    Ok(mesh)
+}
+
+async fn maybe_find_peer(role: Option<&ServalRole>, override_var: &str) -> Result<SocketAddr> {
+    if let Ok(override_url) = std::env::var(override_var) {
+        if let Ok(override_addr) = override_url.parse::<SocketAddr>() {
+            return Ok(override_addr);
+        }
+    }
+
+    log::info!("Looking for {role:?} node on the peer network...");
```

Comment:
Ok, so this is a super interesting function! I have been noodling on this problem myself, and it's how I came up with the "give me a peer with this role" HTTP endpoint.

I was thinking of doing something like this (in terrible pseudocode):

````rs
async fn maybe_find_peer(role: Option<&ServalRole>, override_var: &str) -> Result<KaboodlePeer> {
    let mesh_member = Kaboodle::discover_mesh_member().await
    let Some(required_role) = role else {
        // hey, no required role! that's great, this peer will do
        return mesh_member;
    }

    let http_addr = format!("http://{}", mesh_member.http_address());
    let path = build_url(format!("/mesh/members/{required_role}?count=1"));
    let url = format!("{http_addr}{path}", 

    //  ...and then fetch `url`, parse the response into a KaboodlePeer, return it
}
````

Reply:
Yeah—I think we should never, ever join the mesh unless we're actually there to participate (or be an observer).